cpp/include/raft/comms/comms.hpp

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+
+namespace raft {
+namespace comms {
+
+typedef unsigned int request_t;
+enum datatype_t { CHAR, UINT8, INT, UINT, INT64, UINT64, FLOAT, DOUBLE };
+enum op_t { SUM, PROD, MIN, MAX };
+
+/**
+ * The resulting status of distributed stream synchronization
+ */
+enum status_t {
+  commStatusSuccess,  // Synchronization successful
+  commStatusError,    // An error occured querying sync status
+  commStatusAbort     // A failure occurred in sync, queued operations aborted
+};
+
+class comms_iface {
+ public:
+  virtual ~comms_iface();
+
+  virtual int getSize() const = 0;
+  virtual int getRank() const = 0;
+
+  virtual std::unique_ptr<comms_iface> commSplit(int color, int key) const = 0;
+
+  virtual void barrier() const = 0;
+
+  virtual status_t syncStream(cudaStream_t stream) const = 0;
+
+  virtual void isend(const void* buf, int size, int dest, int tag,
+                     request_t* request) const = 0;
+
+  virtual void irecv(void* buf, int size, int source, int tag,
+                     request_t* request) const = 0;
+
+  virtual void waitall(int count, request_t array_of_requests[]) const = 0;
+
+  virtual void allreduce(const void* sendbuff, void* recvbuff, int count,
+                         datatype_t datatype, op_t op,
+                         cudaStream_t stream) const = 0;
+
+  virtual void bcast(void* buff, int count, datatype_t datatype, int root,
+                     cudaStream_t stream) const = 0;
+
+  virtual void reduce(const void* sendbuff, void* recvbuff, int count,
+                      datatype_t datatype, op_t op, int root,
+                      cudaStream_t stream) const = 0;
+
+  virtual void allgather(const void* sendbuff, void* recvbuff, int sendcount,
+                         datatype_t datatype, cudaStream_t stream) const = 0;
+
+  virtual void allgatherv(const void* sendbuf, void* recvbuf,
+                          const int recvcounts[], const int displs[],
+                          datatype_t datatype, cudaStream_t stream) const = 0;
+
+  virtual void reducescatter(const void* sendbuff, void* recvbuff,
+                             int recvcount, datatype_t datatype, op_t op,
+                             cudaStream_t stream) const = 0;
```

Comment:
the assumption, at least on ML, is that these should always be CUDA Pointers and we should be doing as little as possible on host.

Reply:
Yes, if the data resides in GPU memory, we should avoid moving data back to host memory.

But say we are doing something like the following.
```
float error_sum = thrust::reduce(...);
comms.reduce(&error_sum, &error_sum, 1);
```
If our communicator does not support host memory, we need to transfer error_sum back to GPU memory, run comms.reduce(), then copy-back to CPU memory to figure out error is small enough or not.

More code, longer latency. I don't think this is desirable.

Reply:
It sounds like there are two different concerns here.  
*More code: * This sounds like a convenience feature and I'm not opposed to it. I'd prefer to open a Github issue and for someone to tackle it after the initial merge, though. What do you think?

*Longer latency: * Unfortunately, this is unavoidable as both UCX and NCCL are going to stage on host when a GPU-based transport is not available (eg. over network w/o infiniband/gpudirect). Wouldn't pinned host memory also reduce the latency here?

Reply:
>  I'd prefer to open a Github issue and for someone to tackle it after the initial merge, though. What do you think?

Sure, no problem.

>  Longer latency: Unfortunately, this is unavoidable as both UCX and NCCL are going to stage on host when a GPU-based transport is not available (eg. over network w/o infiniband/gpudirect). Wouldn't pinned host memory also reduce the latency here?

Oh, I am not talking about data residing in GPU memory and using host memory as a temporary buffer. I am talking about data actually residing in host memory (e.g. thrust reduce function returns reduced value in host memory). If we are communicating data residing in host memory, no need to involve GPU, and it will be nice if communicator supports this.
```
template <typename value_t>
void Environment::reduce(
    Communicator comm, Target target,
    value_t* p_src, value_t* p_dst, size_t value_count, int root, ReduceOp reduce_op,
    cudaStream_t stream) {
  MPI_Comm* p_mpi_comm{nullptr};
  ncclComm_t* p_nccl_comm{nullptr};
  std::tie(p_mpi_comm, p_nccl_comm) = get_comm_ptrs(comm);

  if (target == Target::HOST) {
    auto mpi_op = get_mpi_reduce_op(reduce_op);
    if (p_src == p_dst) {
      int comm_rank{};
      MPI_TRY(MPI_Comm_rank(*p_mpi_comm, &comm_rank));
      if (comm_rank == root) {
        MPI_TRY(
          MPI_Reduce(
            MPI_IN_PLACE, p_dst, value_count, get_mpi_type<value_t>(), mpi_op, root, *p_mpi_comm));
      }
      else {
        MPI_TRY(
          MPI_Reduce(
            p_src, p_dst, value_count, get_mpi_type<value_t>(), mpi_op, root, *p_mpi_comm));
      }
    }
    else {
      MPI_TRY(
        MPI_Reduce(p_src, p_dst, value_count, get_mpi_type<value_t>(), mpi_op, root, *p_mpi_comm));
    }
  }
  else if (target == Target::DEVICE) {
    auto nccl_op = get_nccl_reduce_op(reduce_op);
    NCCL_TRY(
      ncclReduce(
       p_src, p_dst, value_count, get_nccl_type<value_t>(), nccl_op, root,
       *p_nccl_comm, stream));
  }
  else {
    CUGRAPH_FAIL("Unsupported target");
  }
}
```
I have done something like the above (and we can move "Target target" as a template parameter to resolve this in compile time") to make communication library functions work for both host memory data and device memory.

No need to implement this in this PR, but we need to eventually add this support (and if we cannot use MPI for python/dask interaction, we need to implement this with UCX, but AFAIK, UCX figures out memory location from address, so we can largely reuse our UCX backend code for both host & device, correct?)

Reply:
That's correct- UCX supports this already but won't provide the collectives. We weren't able to get MPI working in Dask because of some of the assumptions it makes with regards to communicator lifecycles within a process. 

I suppose one option could be for us to create a feature request with the NCCL project. Not sure if we want to re-implement ring- and tree-style reductions w/ UCX, though I suppose that's another option.

Reply:
Reimplementing collectives using UCX sounds like an insane amount of work for us and largely repeating what MPI folks already did, better avoid unless there is really no other way.

It will be very nice if NCCL supports this :-)